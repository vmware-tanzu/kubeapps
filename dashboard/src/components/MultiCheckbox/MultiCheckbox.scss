// Copyright 2021-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

.multicheckbox-wrapper {
  div {
    width: fit-content;
    margin-top: 0.5em;
  }
}
