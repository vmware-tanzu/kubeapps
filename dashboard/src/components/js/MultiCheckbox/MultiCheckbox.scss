// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

// This breakpoints matches the ones stated in Clarity at: https://v2.clarity.design/grid

// small
@media (min-width: 576px) {
  .multicheckbox-wrapper {
    columns: var(--col-sm-num);
  }
}

// medium
@media (min-width: 768px) {
  .multicheckbox-wrapper {
    columns: var(--col-md-num);
  }
}

// large
@media (min-width: 992px) {
  .multicheckbox-wrapper {
    columns: var(--col-lg-num);
  }
}

// extra large
@media (min-width: 1200px) {
  .multicheckbox-wrapper {
    columns: var(--col-xl-num);
  }
}
