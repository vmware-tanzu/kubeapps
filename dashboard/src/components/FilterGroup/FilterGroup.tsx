// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import MultiCheckbox from "components/MultiCheckbox";
import React from "react";

interface IFilterGroupProps {
  name: string;
  options: string[];
  currentFilters: string[];
  onAddFilter: (type: string, value: string) => void;
  onRemoveFilter: (etype: string, value: string) => void;
  disabled?: boolean;
}

function FilterGroup({
  name,
  currentFilters,
  options,
  disabled = false,
  onAddFilter,
  onRemoveFilter,
}: IFilterGroupProps) {
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value: targetValue } = e.target;
    const index = currentFilters.indexOf(targetValue);
    if (index > -1) {
      // Already checked, remove
      onRemoveFilter(name, targetValue);
    } else {
      // Not checked, add
      onAddFilter(name, targetValue);
    }
  };
  return (
    <MultiCheckbox
      name={name}
      options={options}
      value={currentFilters}
      onChange={onChange}
      disabled={disabled}
    />
  );
}

export default FilterGroup;
