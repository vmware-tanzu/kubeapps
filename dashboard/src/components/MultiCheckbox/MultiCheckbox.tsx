// Copyright 2021-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { CdsCheckbox } from "@cds/react/checkbox/";
import "./MultiCheckbox.scss";

export interface IMultiCheckboxProps {
  name: string;
  value: string[];
  options: string[];
  disabled?: boolean;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export function MultiCheckbox({ name, value, options, disabled, onChange }: IMultiCheckboxProps) {
  return (
    <div className="multicheckbox-wrapper">
      {options.map((opt, i) => (
        <div key={i}>
          <CdsCheckbox>
            <label htmlFor={`${name}-${i}`}>{opt}</label>
            <input
              type="checkbox"
              value={opt}
              id={`${name}-${i}`}
              checked={value.includes(opt)}
              disabled={disabled}
              onChange={onChange}
            />
          </CdsCheckbox>
        </div>
      ))}
    </div>
  );
}
