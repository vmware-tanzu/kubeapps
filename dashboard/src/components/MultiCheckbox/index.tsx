// Copyright 2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { MultiCheckbox } from "./MultiCheckbox";

export default MultiCheckbox;
