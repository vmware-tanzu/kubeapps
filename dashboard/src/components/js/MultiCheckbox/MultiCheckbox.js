// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import PropTypes from "prop-types";
import React from "react";
import "./MultiCheckbox.scss";
import { assignCssVariables } from "./utils/cssVariables";

const MultiCheckbox = ({ options, span, value, ...otherProps }) => (
  <div className="multicheckbox-wrapper" style={assignCssVariables(span)}>
    {options.map((opt, i) => (
      <div key={i} className="clr-checkbox-wrapper">
        <input
          {...otherProps}
          type="checkbox"
          value={opt}
          id={`${otherProps.name}-${i}`}
          checked={value.includes(opt)}
        />
        <label className="clr-control-label" htmlFor={`${otherProps.name}-${i}`}>
          {opt}
        </label>
      </div>
    ))}
  </div>
);

MultiCheckbox.propTypes = {
  span: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.number), PropTypes.number]),
  value: PropTypes.array,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default MultiCheckbox;
