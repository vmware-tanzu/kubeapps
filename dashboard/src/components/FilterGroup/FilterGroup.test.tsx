// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import MultiCheckbox from "components/MultiCheckbox";
import { act } from "react-dom/test-utils";
import { defaultStore, mountWrapper } from "shared/specs/mountWrapper";
import FilterGroup from "./FilterGroup";

const defaultProps = {
  name: "test",
  options: ["foo", "bar"],
  currentFilters: [],
  disabled: false,
  onAddFilter: jest.fn(),
  onRemoveFilter: jest.fn(),
};

it("renders a multicheckbox", () => {
  const wrapper = mountWrapper(defaultStore, <FilterGroup {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

it("calls onChange function", () => {
  const currentFilters: string[] = [];
  const onAddFilter = jest.fn((_n, f) => currentFilters.push(f));
  const onRemoveFilter = jest.fn();
  const wrapper = mountWrapper(
    defaultStore,
    <FilterGroup
      {...defaultProps}
      currentFilters={currentFilters}
      onAddFilter={onAddFilter}
      onRemoveFilter={onRemoveFilter}
    />,
  );
  act(() => {
    (wrapper.find(MultiCheckbox).prop("onChange") as any)({ target: { value: "foo" } });
  });
  expect(onAddFilter).toHaveBeenCalledWith("test", "foo");
  // Force re-render
  wrapper.setProps({ ...defaultProps, onAddFilter, onRemoveFilter });
  // Adds a new item to the filter
  act(() => {
    (wrapper.find(MultiCheckbox).prop("onChange") as any)({ target: { value: "bar" } });
  });
  expect(onAddFilter).toHaveBeenCalledWith("test", "bar");
  // Force re-render
  wrapper.setProps({ ...defaultProps, onAddFilter, onRemoveFilter });
  // Removes an item
  act(() => {
    (wrapper.find(MultiCheckbox).prop("onChange") as any)({ target: { value: "foo" } });
  });
  expect(onRemoveFilter).toHaveBeenCalledWith("test", "foo");
});
