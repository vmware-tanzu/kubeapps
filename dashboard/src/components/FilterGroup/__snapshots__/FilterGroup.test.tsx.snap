// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a multicheckbox 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <IntlProvider
    defaultFormats={Object {}}
    defaultLocale="en"
    fallbackOnEmptyString={true}
    formats={Object {}}
    key="en"
    locale="en"
    messages={
      Object {
        "Kubeapps": Array [
          Object {
            "type": 0,
            "value": "Kubeapps",
          },
        ],
        "Submit": Array [
          Object {
            "type": 0,
            "value": "Submit",
          },
        ],
        "Token": Array [
          Object {
            "type": 0,
            "value": "Token",
          },
        ],
        "error-login-token": Array [
          Object {
            "type": 0,
            "value": "There was an error connecting to the Kubernetes API. Please check that your token is valid.",
          },
        ],
        "login-desc-oidc": Array [
          Object {
            "type": 0,
            "value": "Your cluster operator has enabled login via an authentication provider.",
          },
        ],
        "login-desc-token": Array [
          Object {
            "type": 0,
            "value": "Your cluster operator should provide you with a Kubernetes API token.",
          },
        ],
        "login-oidc": Array [
          Object {
            "type": 0,
            "value": "Login via OIDC Provider",
          },
        ],
        "login-title-welcome": Array [
          Object {
            "type": 0,
            "value": "Welcome to",
          },
        ],
        "more-info": Array [
          Object {
            "type": 0,
            "value": "More Info",
          },
        ],
        "paste-token-here": Array [
          Object {
            "type": 0,
            "value": "Paste token here",
          },
        ],
      }
    }
    onError={[Function]}
    onWarn={[Function]}
    textComponent={Symbol(react.fragment)}
  >
    <BrowserRouter>
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <FilterGroup
          currentFilters={Array []}
          disabled={false}
          name="test"
          onAddFilter={[MockFunction]}
          onRemoveFilter={[MockFunction]}
          options={
            Array [
              "foo",
              "bar",
            ]
          }
        >
          <MultiCheckbox
            disabled={false}
            name="test"
            onChange={[Function]}
            options={
              Array [
                "foo",
                "bar",
              ]
            }
            value={Array []}
          >
            <div
              className="multicheckbox-wrapper"
            >
              <div
                key="0"
              >
                <CdsCheckbox>
                  <CdsCheckbox
                    _$Gl={null}
                  >
                    <cds-checkbox
                      _$Gl={null}
                    >
                      <label
                        htmlFor="test-0"
                      >
                        foo
                      </label>
                      <input
                        checked={false}
                        disabled={false}
                        id="test-0"
                        onChange={[Function]}
                        type="checkbox"
                        value="foo"
                      />
                    </cds-checkbox>
                  </CdsCheckbox>
                </CdsCheckbox>
              </div>
              <div
                key="1"
              >
                <CdsCheckbox>
                  <CdsCheckbox
                    _$Gl={null}
                  >
                    <cds-checkbox
                      _$Gl={null}
                    >
                      <label
                        htmlFor="test-1"
                      >
                        bar
                      </label>
                      <input
                        checked={false}
                        disabled={false}
                        id="test-1"
                        onChange={[Function]}
                        type="checkbox"
                        value="bar"
                      />
                    </cds-checkbox>
                  </CdsCheckbox>
                </CdsCheckbox>
              </div>
            </div>
          </MultiCheckbox>
        </FilterGroup>
      </Router>
    </BrowserRouter>
  </IntlProvider>
  ,
</Provider>
`;
